// ------------------------------------------------------------------
//   sam_gem3.c
//   Copyright (C) 2022-2025 Genozip Limited. Patent pending.
//   Please see terms and conditions in the file LICENSE.txt
//
//   WARNING: Genozip is proprietary, not open source software. Modifying the source code is strictly prohibited,
//   under penalties specified in the license.

// Compresses auxilliary fields generated by hisat2 mapper

#include "sam_private.h"

// HISAT2 Zs:Z field: "When the alignment of a read involves SNPs that are in the index, this option is used to indicate where exactly the read 
//                     involves the SNPs. This optional field is similar to the above MD:Z field. For example, Zs:Z:1|S|rs3747203,97|S|rs16990981 
//                     indicates the second base of the read corresponds to a known SNP (ID: rs3747203). 97 bases after the third base (the base 
//                     after the second one), the read at 100th base involves another known SNP (ID: rs16990981). ‘S’ indicates a single nucleotide 
//                     polymorphism. ‘D’ and ‘I’ indicate a deletion and an insertion, respectively." (http://daehwankimlab.github.io/hisat2/manual/)
void sam_seg_HISAT2_Zs_Z (VBlockSAMP vb, STRp(zs), unsigned add_bytes)
{
    static const MediumContainer con = {
        .nitems_lo   = 3, 
        .drop_final_repsep = true,
        .repsep      = {','},
        .items       = { { .dict_id.num = _OPTION_Zs_POS,  .separator = {'|'} },
                         { .dict_id.num = _OPTION_Zs_TYPE, .separator = {'|'} },
                         { .dict_id.num = _OPTION_Zs_RS                       } }
    };
    
    seg_array_of_struct (VB, CTX(OPTION_Zs_Z), con, STRa(zs), (SegCallback[]){0, 0, seg_id_field_varlen_int_cb}, NULL, add_bytes);
}

